### ISSUE TITLE
Specifying substrings

### ISSUE THREAD

**Issue by xrotwang** [cldf/cldf#122]
@cysouw proposes

> would it be an idea to use a concept like Substring in general throughout CLDF to refer to any part of a Form by counting unicode symbols (or something like that)? If you think this would be feasible, then we should open a different issue for that :)

See https://github.com/cldf/cldf/issues/120#issuecomment-1040318973

**Comment by xrotwang** [comment]
Maybe we could simply start with a recommendation how to specify substring references in CLDF columns. I think the [Segment_Slice](https://cldf.clld.org/v1.0/terms.rdf#segmentSlice) columns spec can be reused here. What's laccking is probably a clear statement that we are talking about strings as "lists of unicode code points" here, not phonemes, graphemes or similar.

**Comment by cysouw** [comment]
The idea would be to allow a `Segment_Slice`-like specification for any substring of any `Form` as listed in any `FormTable`. In that way anybody has a method to refer to any selection of symbols in the data.

The problem I see indeed is how to count characters. Especially with diacritics and the like this might not be easy for 'normal' users ... One would have to be sure about Unicode Normalizations and the like.

**Comment by cysouw** [comment]
Maybe the easiest way would be to suggest to everybody to always add a column like `Characters` to a `FormTable`, meaning that the provider of the data suggest some kind of 'minimal' separation into characters. This columns `Characters` would then again be space-separated and can easily be referenced.

That would technically be the easiest solution, but it would result in some extra work for people preparing data.

**Comment by xrotwang** [comment]
Yes, specifying this technically isn't too hard. (Here's the reference implementation https://github.com/cldf/pycldf/blob/ee7a8028c6792f946e1995f3eec8a38b84289a36/src/pycldf/util.py#L42-L54 - which I had to lookup to figure out whether indices are supposed to be 0 or 1-based :) ).

But as you say, specifying substring indices would be non-trivial, or rather impossible to do by hand, just looking at the data.

**Comment by xrotwang** [comment]
And once substring indices are computed somehow - the same computation would probably also yield the "segmented" data as first step. Storing this first step explicitly would probably make most sense then.

**Comment by cysouw** [comment]
hmmm, the more I think about it, this all suggests that we need different kinds of 'segmentations' for the ontology after all. Currently we use different notions of 'segments' for cognates and for parallel texts. And we will get a new one for morphemes, and now maybe another one for characters...

Should we maybe just specify them as such?

- Character_Segments
- Sound_Segments (as used for Cognacy)
- Morph_Segments
- Word_Segments (as used in Parallel texts)

With more kinds of segments being added in the future, depending on the use-cases that people have?

**Comment by xrotwang** [comment]
Sounds reasonable

**Comment by LinguList** [comment]
That would call for some hierarchy, you have the generic Segments, and then some subclasses, like Sound_Segments. In fact, to be a bit more precise, we decided to just call the `Sound_Segments` `sounds` in the Python code of [cltoolkit](https://github.com/cldf/cltoolkit).

**Comment by cysouw** [comment]
@LinguList agree: we leave the generic `Segments` just like it is now, which can be used for whatever segmentation you like to add. When you want to become more precise (or when you want to add more than one segmentation) then we provide the subclasses.

@xrotwang you know more about ontologies: do you like that approach as well?

**Comment by xrotwang** [comment]
I wouldn't try to come up with something similar to a class hierarchy here, but just a list of properties that says "more specific than just Segments" in the description. And I would only add those to the CLDF spec that have well understood applications. So `Character_Segments` wouldn't qualify from my POV - unless someone wanted to model stuff like time-aligned transcribed sound in CLDF,  and needed to reference transcription chunks ...

**Comment by LinguList** [comment]
Not only to the user, but to the application which the users use. If you load your CLDF Wordlist with cltoolkit, you have a very clear interpretation of Segments that assumes a space-segmented string that can have additional morpheme boundaries. There is a bunch of applications tight to it (as shown in our [paper under review](https://doi.org/10.21203/rs.3.rs-870835/v1)), and work on this gave also rise to [cldfviz](https://github.com/cldf/cldfviz), since we had these plots of the data for the paper, and decided to make them more reusable to other applications. It always stands and falls with applications that make use of the data, and how they are maintained and propagated, I'd say.

**Comment by cysouw** [comment]
A problem came up in another discussion and I just note it here for future reference: when you have multiple different `Segment` columns, then it becomes a problem to use `Segment_Slice` because you don't know to which column it refers. That would mean to have additional specifications like `Morph_Segment_Slice` as well...

**Comment by cysouw** [comment]
Another approach (as used by @LinguList) is to allow for boundary symbols in the `Segment` parsing. We could think about this using the regular Leipzig Glossing Rules conventions, with an alternative `+` for morpheme boundaries (for backward compatibility with @LinguList) and a `#` for word boundaries (as used in many approaches).

**Comment by LinguList** [comment]
This would break it down to defining boundary symbols for secondary (morpheme) and tertiary (word) segmentation. As far as I understand, this is already the practice in pyigt, and in applications like the [TPPSR](https://tppsr.clld.org), we have usecases of lexicon (also segmented by morphology), which is linked up with example sentences. Here sentences aren't spelled out as segments, but they could, as they are fully linked.

**Comment by cysouw** [comment]
The Leipzig Glossing Rules go further (besides using `-` for morphemes), and has symbols for other kinds of boundaries, e.g. for clitics `=`, reduplication `~` and infixes `< >`. Some way to notate non-linear morphology is necessary as well (ablaut, umlaut, harmony, tone change, etc). And of course suppletion...

In general: to catch morphological structure in a simple format is not that easy :-).

**Comment by xrotwang** [comment]
For me, one of the guiding principles of CLDF is that only things that have clear, automated applications can be specified. So there's no need to come up with a complete format for morphological structure. OTOH it would be ok to specify even a "dumbed down" subset of this structure in case the appications exist.

And while I think the LGR should be a reference, I think some of its recommendations aren't well suited for a automatically parseable format (e.g. rule 2A, or the infix notation).

**Comment by LinguList** [comment]
There was a lengthy discussion about LGR on LingTyp, specifically the = is ambiguous, as I tried to make clear in that discussion, since it is not clear where to attach an element. So I follow @xrotwang in yes taking LGR as something to mention, but in our [study on IGT with CLDF](https://doi.org/10.1145/3389010), for example, we deliberately discarded the separators that bear semantics over adjacent parts (we had a discussion with @fmatter about this a while ago, if I remember properly).

**Comment by cysouw** [comment]
Then we could interpret `+` as standing for a superset of just any word-internal boundary :-). Suitable, as it is not used by LGR.

(sorry, but I didn't know about your IGT Paper)

**Comment by xrotwang** [comment]
Well, I think in the absence of clear-cut analysis tools/methods, I'd just keep LGR text/gloss **as is** in some non-CLDF-standard column if round-tripping is important. So far, I haven't seen much automated analysis that goes beyond "squeeze out the gloss abbreviations", and that's feasible with the current, mostly-punctuation-agnostic approach.

So "accomodated by the CLDF ecosystem" means
- either there's some tool that expects data with fixed semantics (then this is a candidate for standardisation)
- otherwise CLDF allows to just "keep data along" in custom tables/columns until methods catch up.

**Comment by xrotwang** [comment]
I think parsing LGR unambiguously is already not entirely trivial (see https://github.com/cldf/pyigt/pull/15). So I'm not sure a lot of automated analysis can be built on the finer points of it.

**Comment by LinguList** [comment]
@fmatter, what I meant is that the grammatical meaning/function/nature of a morpheme is determined by the separator, not by an independent column which tells you about morpheme. If you distinguish clitics from bound morphemes, or whatever, but separate something as ba=be, the = notation is ambiguous, since you don't know if ba or be is a clitic. So there is an inherent design flaw in LGR or this kind of notation. The flaw is that = is not just a separator, but transfers meaning to the following or preceding morpheme, in classifying it as a clitic. This is bad annotation style.

**Comment by LinguList** [comment]
https://github.com/cldf/pyigt/issues/6

**Comment by LinguList** [comment]
Above is the issue where we discussed this.

**Comment by fmatter** [comment]
> and we gave up distinguishing between words and morphemes there, since this overcomplicated all analysis, but implicit word boundaries are given just by the Form, that typically corresponds to a word as unit.

since a Form is indeed a WordForm, let me ask the question of "what is a word"? The current implementation in CLDF is the `FormTable`, but what kind of word are we talking about here? If a `Form` is a g-word (or some other morphosyntactically based unit), `=` as used in the LGR is not needed. If it can be something longer, like *The cats are not on the mat.* ([Haspelmath 2020](https://link.springer.com/article/10.1007/s11525-020-09355-5)), then `=` becomes necessary.

As for interlinear sentences, the widespread use of `-`, ` ` and `=` describes *some sort of* morphosyntactic and prosodic unit boundaries, although what those units will differ between languages (and analyses).

**Comment by LinguList** [comment]
In historical linguistics, where you want to make a reconstruction of a language family, the first part of the analysis of identifying words has most time already been done. So the forms that appear in the FormTable are rarely phrases (some exceptions are cases line *estar parado/sentado* for "stand/sit" in Spanish but also French, etc., which even appears as a concept in the Swadesh list). So one can get away without defining words here, or say: I work with the unit that expresses my concept list.

I wonder how large the number of problematic cases would be out there, which would require to fight for word definitions? If it is two items out of 100, is it worth to keep it in mind when modeling something? (also thinking loudly without having finished the thought)

**Comment by fmatter** [comment]
My current approach assumes that for interlinearized sentences, `\t` and `=` separate the constituents, and uses a separate column `Morpho_Segments` in the [wordform](https://github.com/fmatter/cldf-ldd/tree/main/src/cldf_ldd/components/wordforms) and [stem](https://github.com/fmatter/cldf-ldd/tree/main/src/cldf_ldd/components/stems) models. These are then referenced via association tables containing an index (and glosses!): [wordforms and morphs](https://github.com/fmatter/cldf-ldd/tree/main/src/cldf_ldd/components/wordformparts), [wordforms and stems](https://github.com/fmatter/cldf-ldd/tree/main/src/cldf_ldd/components/formstems), [stems and morphs](https://github.com/fmatter/cldf-ldd/tree/main/src/cldf_ldd/components/stemparts).

* sentence with wordform links, wordforms with morph and gloss links:

![Screenshot from 2023-02-27 16-07-35](https://user-images.githubusercontent.com/2378389/221717931-edfbd393-9ca0-46fb-8d12-4a4e5fcedb68.png)

* morphologically complex stems:

<img src="https://user-images.githubusercontent.com/2378389/221693470-9c811e78-8fef-45be-a648-fb73c6314dc7.png" width="70%" height="70%">



Also, I ran into this problem mentioned by @cysouw:

> A problem came up in another discussion and I just note it here for future reference: when you have multiple different Segment columns, then it becomes a problem to use Segment_Slice because you don't know to which column it refers. That would mean to have additional specifications like Morph_Segment_Slice as well...

**Comment by LinguList** [comment]
Looks cool!

**Comment by xrotwang** [comment]
Hm. I have no ready answer to the question of how to relate columns other than through some sort of naming conventions - which would defeat the purpose of using propertyUrls as abstraction layer.

**Comment by xrotwang** [comment]
Maybe rather than some sort of index notation, the relevant parts of the referenced data should be copied? The advantage of just referencing - i.e. curating the "real" data in just one place - isn't very robust anyway. If, say, morphemes in a word are re-arranged, the indexes will break.

**Comment by fmatter** [comment]
Wordformparts need to have an index, or there needs to be some other way of knowing the order of morphs in a wordform. But yeah, wordformparts could also have information about what they look like -- in fact, you would then not even need a Morpho_Segments column, since the segments would be stored in the wordformparts...
